caddytls/certificate_transparency.go

```diff
@@ -0,0 +1,395 @@
+		sct *signedCertificateTimestamp
+		err error
+	}
+	var wg sync.WaitGroup
+	// Buffered chan so that the goroutines can just exit once they've got a
+	// result.
+	results := make(chan result, len(logs))
+	var submit = func(log ctLog) {
+		defer wg.Done()
+		sct, err := submitSCT(log.url, payload)
+		results <- result{log, sct, err}
+	}
+	wg.Add(len(logs))
+	for _, ctLog := range logs {
+		go submit(ctLog)
+	}
+	wg.Wait()
+	close(results)
+
+	// Chrome CT policy requires at least 1 Google log and 1 non-Google log
+	needGoogle := true
+	needNonGoogle := true
+	for result := range results {
+		if result.err != nil {
+			// Ignore HTTP 4xx errors, which generally indicate "this log isn't
+			// accepting new submissions" (we still want to submit in case they
+			// have a previous SCT for us) or "this log doesn't acecpt certs
+			// from this root"
+			if err, ok := result.err.(*httpResponseError); ok {
+				if err.statusCode >= 400 && err.statusCode < 500 {
+					continue
+				}
+			}
+			log.Printf("[WARNING] Error submitting to CT log: %v", result.err)
+			continue
+		}
+		// Skip logs that don't contribute to our needs.
+		if (result.log.isGoogle && !needGoogle) || (!result.log.isGoogle && !needNonGoogle) {
+			continue
+		}
+		bytes, err := result.sct.AsRawBytes()
+		if err != nil {
+			log.Printf("[WARNING] Got SCT which couldn't be converted to bytes: %v", err)
+			continue
+		}
+		sctBytes = append(sctBytes, bytes)
+		if result.log.isGoogle {
+			needGoogle = false
+		} else {
+			needNonGoogle = false
+		}
+		if !needGoogle && !needNonGoogle {
+			break
+		}
+	}
+	return sctBytes, nil
+}
+
+type ctLog struct {
+	url   string
+	logID [sha256.Size]byte
+	// Chrome's CT policy requires one SCT from a Google log, and one SCT from
+	// a non-Google log, so we track whether a log is Google or not.
+	isGoogle bool
+}
+
+type logList struct {
+	Logs []struct {
+		Description    string `json:"description"`
+		URL            string `json:"url"`
+		Key            string `json:"key"`
+		OperatedBy     []int  `json:"operated_by"`
+		DisqualifiedAt *int   `json:"disqualified_at"`
+	} `json:"logs"`
+	Operators []struct {
+		Name string `json:"name"`
+		ID   int    `json:"id"`
+	} `json:"operators"`
+}
+
+// getTrustedCTLogs returns a list of CT logs trusted by Chrome. As the
+// browser/CT ecosystem evolves it may return other CT logs as well.
+func getTrustedCTLogs() ([]ctLog, error) {
+	response, err := httpClient.Get("https://www.gstatic.com/ct/log_list/log_list.json")
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("HTTP response error: %d", response.StatusCode)
+	}
+	list := logList{}
+	// Limit the amount we read to 10MB to defend against a DoS. Currently this
+	// file is ~2.5KB so this leaves us plenty of breathing room.
+	err = json.NewDecoder(io.LimitReader(response.Body, 10*1024*1024)).Decode(&list)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
What if we wrap the code from the start of this function to this point in another function that loads the cached `logList` or else gets a new one, if the cached one is old. That way we just call `list, err := getLogList()` and most of the time it won't have the overhead of a network request.